velox/functions/prestosql/aggregates/ProductAggregate.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+
+namespace facebook::velox::aggregate::prestosql {
+namespace {
+
+template <typename T, typename U>
+class ProductAggregate {
+ public:
+  using InputType = Row<T>;
+
+  using IntermediateType = U;
+
+  using OutputType = U;
+
+  struct AccumulatorType {
+    U product_{1};
+
+    AccumulatorType() = delete;
+
+    explicit AccumulatorType(HashStringAllocator* /*allocator*/) {}
+
+    void addInput(HashStringAllocator* /*allocator*/, exec::arg_type<T> value) {
+      product_ *= value;
+    }
+
+    void combine(HashStringAllocator* /*allocator*/, exec::arg_type<U> other) {
+      product_ *= other;
+    }
+
+    bool writeFinalResult(exec::out_type<OutputType>& out) {
+      out = product_;
+      return true;
+    }
+
+    bool writeIntermediateResult(exec::out_type<IntermediateType>& out) {
+      out = product_;
+      return true;
+    }
+  };
+
+  static bool toIntermediate(exec::out_type<U>& out, exec::arg_type<T> in) {
```

Comment:
nit: shall we put input first?

const exec::arg_type<T>& in? Or T is always scaler type? Thanks!

Reply:
> shall we put input first?

This API is part of the framework, hence, cannot just in the function.

> Or T is always scaler type?

T is either integer or floating point number.